server/conn_stmt.go

```diff
@@ -98,8 +98,10 @@ func (cc *clientConn) handleStmtPrepare(ctx context.Context, sql string) error {
 			}
 		}
 
-		if err := cc.writeEOF(0); err != nil {
-			return err
+		if cc.capability&mysql.ClientDeprecateEOF == 0 {
```

Comment:
Hi, according to the MySQL doc https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_com_stmt_prepare.html#sect_protocol_com_stmt_prepare_response.

> if num_params > 0 and CLIENT_OPTIONAL_RESULTSET_METADATA is not set or if medatdata_follows is RESULTSET_METADATA_FULL num_params packets will follow. Then a [EOF_Packet](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_eof_packet.html) will be transmitted, provided that CLIENT_DEPRECATE_EOF is not set.

It seems we always send num_params packets? why can this work?

Reply:
There is no `CLIENT_OPTIONAL_RESULTSET_METADATA` defined in TiDB yet.

https://github.com/pingcap/tidb/blob/37bd052ee53d1180a89e12b6a01a2ec854f357b6/parser/mysq/const.go#L163

I think the second sentence in that quote is not dependent on the first part.

So maybe read it like:

> if num_params > 0 and CLIENT_OPTIONAL_RESULTSET_METADATA is not set or if medatdata_follows is RESULTSET_METADATA_FULL num_params packets will follow.

> Then a [EOF_Packet](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_eof_packet.html) will be transmitted, provided that CLIENT_DEPRECATE_EOF is not set.


And when we exclude `CLIENT_OPTIONAL_RESULTSET_METADATA` and `medatdata_follows` (which requires the former) then it is like this:

> if num_params > 0, num_params packets will follow.

> Then a [EOF_Packet](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_eof_packet.html) will be transmitted, provided that CLIENT_DEPRECATE_EOF is not set.

This seems to match what the code does.

Would be good to compare this with MySQL packet traces if you think MySQL is behaving differently.